[OpenMP][OMPIRBuilder] Handle non-failing calls properly

skatrak opened the pull request:
The preprocessor definition used to enable asserts and the one that `llvm::Error` and `llvm::Expected` use to ensure all created instances are checked are not the same. By making these checks inside of an `assert` in cases where errors are not expected, certain build configurations would trigger runtime failures (e.g. `-DLLVM_ENABLE_ASSERTIONS=OFF -DLLVM_UNREACHABLE_OPTIMIZE=ON`).

The `llvm::cantFail()` function, which was intended for this use case, is used by this patch in place of `assert` to prevent these runtime failures. For tests, the `llvm::expectedToOptional()` in combination with `ASSERT_TRUE` and `std::optional`'s builtin asserts are used instead, to avoid silent failures in release builds.

skatrak committed:
[OpenMP][OMPIRBuilder] Handle non-failing calls properly

The preprocessor definition used to enable asserts and the one that
`llvm::Error` and `llvm::Expected` use to ensure all created instances are
checked are not the same. By making these checks inside of an `assert` in cases
where errors are not expected, certain build configurations would trigger
runtime failures (e.g. `-DLLVM_ENABLE_ASSERTIONS=OFF
-DLLVM_UNREACHABLE_OPTIMIZE=ON`).

The `llvm::cantFail()` function, which was intended for this use case, is used
by this patch in place of `assert` to prevent these runtime failures. In tests,
new preprocessor definitions based on `ASSERT_THAT_EXPECTED` and
`EXPECT_THAT_EXPECTED` are used instead, to avoid silent failures in release
builds.

Meinersbur commented:
`cantFail` is not guaranteed to crash, e.g. in `-DLLVM_ENABLE_ASSERTIONS=OFF -DLLVM_UNREACHABLE_OPTIMIZE=ON`. I think this is fine for implementation code, for which `LLVM_UNREACHABLE_OPTIMIZE` is meant to improve performance, but for unittests such tests could silently pass in release builds. 

Ideas:
 * `ASSERT_FALSE(errorToBool(Err))`
 * `if (Err) report_fatal_error(Err)`
 * Some new helper function

skatrak commented:
Thank you @Meinersbur for your comments. I just updated the PR description and replaced `cantFail()` calls in unit tests as you suggested, let me know if that addresses your concerns.

Meinersbur commented on llvm/unittests/Frontend/OpenMPIRBuilderTest.cpp:
```diff
@@ -345,18 +344,15 @@ TEST_F(OpenMPIRBuilderTest, CreateBarrier) {
-  OpenMPIRBuilder::InsertPointOrErrorTy BarrierIP1 =
-      OMPBuilder.createBarrier({IRBuilder<>::InsertPoint()}, OMPD_for);
-  assert(BarrierIP1 && "unexpected error");
+  ASSERT_TRUE(expectedToOptional(
+      OMPBuilder.createBarrier({IRBuilder<>::InsertPoint()}, OMPD_for)));
```
`llvm::Expected` as well as `std::optional` are convertable to `bool`, but because `ASSERT_TRUE` actually converts to `AssertionResult`, only the latter works. So this is the shorter version of the same (which I have been using so far):
```suggestion
  ASSERT_TRUE((bool)OMPBuilder.createBarrier({IRBuilder<>::InsertPoint()}, OMPD_for));
```

I hoped you could find something nicer. Some tests define their own macro `ASSERT_NO_ERROR`. LLVM itself defines `ASSERT_THAT_EXPECTED` to be used in GTest. None of these return the value in-case of no-error, so ugly to use, but seem to be the canonical way to do the check. How about a helper function similar to `cantFail`, but guaranteed to crash instead of undefined behaviour?
```
template<typename T>
T successOrAbort(Expected<T> &&ValOrErr) {
    if (ValOrErr)
      return std::move(*ValOrErr);
  report_fatal_error(ValOrErr.takeError());
}
```

Something really cool would properly integreate into the GTest envorinment. 
```
template<typename T>
T successOrAbort(Expected<T> &&ValOrErr) {
    if (ValOrErr)
      return std::move(*ValOrErr);
  GTEST_NONFATAL_FAILURE_("Expected<T> contains error value"); // Might be considered internal to GTest.
  abort(); // Cannot continue without a value to unwrap
}
```
or make it a macro such we can customize the  message  and use `GTEST_MESSAGE_AT_` and report the affected line number.

skatrak replied:
I made another revision to avoid translating to `std::optional` in order to discard errors, and instead handle them through some new local macros. I'm not against adding these functions that you propose, but that doesn't seem like a good thing to do as part of this PR. Perhaps something that can be added separately and then potentially update this and other tests to use it.

Meinersbur reviewed (approved):
The compromise with `cantFail` is fine. Most will run those tests with assertions enabled and undefined behaviour is one of the things a unittests should be regularly tested for (e.g. with undefined-behaviour sanitizer).

skatrak committed:
Merge branch 'main' into ompirbuilder-err-assert
